Compilation Error when using Any-Matcher with Eigen::Vector2f argument

Below code gets an error message when compiling with g++ 7.3.0:

```
#include "catch.hpp"
#include "fakeit.hpp"

#include "Eigen/Dense" // from http://eigen.tuxfamily.org/index.php?title=Main_Page
using Eigen::Vector2f;

struct S {
int a; 
int b;
};

class C3 {
public:
  virtual void f(Vector2f s1) = 0;
};


TEST_CASE("Fakeit cannot resolve overload with Vector2f when using any-matcher?") {
  using namespace fakeit;
  Mock<C3> mock;

  Verify(Method(mock, f)
  	 .Using(_)).Once();
}
```

The error message is:
```
/usr/bin/c++ -Wall -Wno-unknown-pragmas -std=c++1z -c -I -Ieigen -I -o test_fakeit.cpp.o /tests/test_fakeit.cpp
In file included from /tests/test_fakeit.cpp:2:0:
/tests/fakeit.hpp: In instantiation of ‘void fakeit::MethodMockingContext<R, arglist>::setMatchingCriteria(const matcherCreators& ...) [with matcherCreators = {fakeit::AnyMatcher}; <template-parameter-2-2> = void; R = void; arglist = {Eigen::Matrix<float, 2, 1, 0, 2, 1>}]’:
/tests/fakeit.hpp:7895:72:   required from ‘fakeit::MockingContext<void, arglist ...>& fakeit::MockingContext<void, arglist ...>::Using(const arg_matcher& ...) [with arg_matcher = {fakeit::AnyMatcher}; arglist = {Eigen::Matrix<float, 2, 1, 0, 2, 1>}]’
/tests/test_fakeit.cpp:71:3:   required from here
/tests/fakeit.hpp:7787:13: error: call of overloaded ‘CollectMatchers(const fakeit::AnyMatcher&)’ is ambiguous
             c.CollectMatchers(matcherCreator...);
             ^
/tests/fakeit.hpp:7529:16: note: candidate: typename std::enable_if<std::is_constructible<fakeit::MatchersCollector<index, arglist>::NakedArgType<index>, Head>::value, void>::type fakeit::MatchersCollector<index, arglist>::CollectMatchers(const Head&) [with Head = fakeit::AnyMatcher; unsigned int index = 0; arglist = {Eigen::Matrix<float, 2, 1, 0, 2, 1>}; typename std::enable_if<std::is_constructible<fakeit::MatchersCollector<index, arglist>::NakedArgType<index>, Head>::value, void>::type = void]
         ::type CollectMatchers(const Head &value) {
                ^~~~~~~~~~~~~~~
/tests/fakeit.hpp:7539:16: note: candidate: typename std::enable_if<std::is_constructible<fakeit::MatchersCollector<index, arglist>::NakedArgType<index>, Head>::value, void>::type fakeit::MatchersCollector<index, arglist>::CollectMatchers(const Head&, const Tail& ...) [with Head = fakeit::AnyMatcher; Tail = {}; unsigned int index = 0; arglist = {Eigen::Matrix<float, 2, 1, 0, 2, 1>}; typename std::enable_if<std::is_constructible<fakeit::MatchersCollector<index, arglist>::NakedArgType<index>, Head>::value, void>::type = void]
         ::type CollectMatchers(const Head &head, const Tail &... tail) {
                ^~~~~~~~~~~~~~~
/tests/fakeit.hpp:7566:16: note: candidate: typename std::enable_if<std::is_same<fakeit::AnyMatcher, Head>::value, void>::type fakeit::MatchersCollector<index, arglist>::CollectMatchers(const Head&) [with Head = fakeit::AnyMatcher; unsigned int index = 0; arglist = {Eigen::Matrix<float, 2, 1, 0, 2, 1>}; typename std::enable_if<std::is_same<fakeit::AnyMatcher, Head>::value, void>::type = void]
         ::type CollectMatchers(const Head &) {
                ^~~~~~~~~~~~~~~
/tests/fakeit.hpp:7574:16: note: candidate: typename std::enable_if<std::is_same<fakeit::AnyMatcher, Head>::value, void>::type fakeit::MatchersCollector<index, arglist>::CollectMatchers(const Head&, const Tail& ...) [with Head = fakeit::AnyMatcher; Tail = {}; unsigned int index = 0; arglist = {Eigen::Matrix<float, 2, 1, 0, 2, 1>}; typename std::enable_if<std::is_same<fakeit::AnyMatcher, Head>::value, void>::type = void]
         ::type CollectMatchers(const Head &head, const Tail &... tail) {
                ^~~~~~~~~~~~~~~

Compilation exited abnormally with code 1 at Tue Mar 12 07:46:25
```

The code compiles, when Vector2f is replaced by a simple struct S:

```
class C3 {
public:
  virtual void f(S s1) = 0;
};
```

Any hints on possible workarounds?

--- a/include/fakeit/MatchersCollector.hpp
+++ b/include/fakeit/MatchersCollector.hpp
@@ -34,6 +34,12 @@ namespace fakeit {
         template<std::size_t N>
         using NakedArgType = typename naked_type<ArgType<index>>::type;
 
+        template <typename MatcherCreatorT, typename = void>
+        struct IsMatcherCreatorTypeCompatible : std::false_type {};
+
+        template <typename MatcherCreatorT>
+        struct IsMatcherCreatorTypeCompatible<MatcherCreatorT, typename std::enable_if<MatcherCreatorT::template IsTypeCompatible<NakedArgType<index>>::value, void>::type> : std::true_type {};
+
         MatchersCollector(std::vector<Destructible *> &matchers)
                 : _matchers(matchers) {
         }
@@ -43,6 +49,8 @@ namespace fakeit {
 
         template<typename Head>
         typename std::enable_if< //
+                !std::is_same<AnyMatcher, typename naked_type<Head>::type>::value &&
+                !IsMatcherCreatorTypeCompatible<typename naked_type<Head>::type>::value &&
                 std::is_constructible<NakedArgType<index>, Head&&>::value, void> //
         ::type CollectMatchers(Head &&value) {
 
@@ -52,7 +60,7 @@ namespace fakeit {
 
         template<typename Head>
         typename std::enable_if< //
-                naked_type<Head>::type::template IsTypeCompatible<NakedArgType<index>>::value, void> //
+                IsMatcherCreatorTypeCompatible<typename naked_type<Head>::type>::value, void> //
         ::type CollectMatchers(Head &&creator) {
             TypedMatcher<NakedArgType<index>> *d = creator.template createMatcher<NakedArgType<index>>();
             _matchers.push_back(d);

--- a/tests/argument_matching_tests.cpp
+++ b/tests/argument_matching_tests.cpp
@@ -25,6 +25,12 @@ namespace {
 	bool operator==(const Base& lhs, const Base& rhs) {
 		return lhs.value() == rhs.value();
 	}
+
+	struct WeirdType
+	{
+		template <typename T>
+		WeirdType(T&&){}
+	};
 }
 
 struct ArgumentMatchingTests: tpunit::TestFixture {
@@ -40,7 +46,7 @@ struct ArgumentMatchingTests: tpunit::TestFixture {
 					TEST(ArgumentMatchingTests::test_str_le_matcher), TEST(ArgumentMatchingTests::test_str_ne_matcher),
 					TEST(ArgumentMatchingTests::test_approx_eq_matcher),
 					TEST(ArgumentMatchingTests::test_any_matcher), TEST(ArgumentMatchingTests::test_any_matcher2),
-					TEST(ArgumentMatchingTests::test_any_matcher3),
+					TEST(ArgumentMatchingTests::test_any_matcher3), TEST(ArgumentMatchingTests::test_any_matcher_weird_constructor),
 					TEST(ArgumentMatchingTests::pass_reference_by_value),
 					TEST(ArgumentMatchingTests::format_Any), TEST(ArgumentMatchingTests::format_Eq),
 					TEST(ArgumentMatchingTests::format_Gt), TEST(ArgumentMatchingTests::format_Ge),
@@ -79,6 +85,7 @@ struct ArgumentMatchingTests: tpunit::TestFixture {
 		virtual int funcMoveOnly(MoveOnlyType) = 0;
 		virtual int funcSlicing(const Base&) = 0;
 		virtual int funcDouble(double) = 0;
+		virtual int funcWeirdType(int, WeirdType) = 0;
     };
 
 	void mixed_matchers() {
@@ -346,6 +353,20 @@ struct ArgumentMatchingTests: tpunit::TestFixture {
 		ASSERT_EQUAL(1, i.func(1));
 
 		Verify(Method(mock, func).Using(Any())).Twice();
+	}
+
+	void test_any_matcher_weird_constructor() {
+		Mock<SomeInterface> mock;
+
+		When(Method(mock, funcWeirdType).Using(1, _)).AlwaysReturn(1);
+		When(Method(mock, funcWeirdType).Using(5, Any())).AlwaysReturn(5);
+
+		SomeInterface &i = mock.get();
+		ASSERT_EQUAL(1, i.funcWeirdType(1, WeirdType{10}));
+		ASSERT_EQUAL(5, i.funcWeirdType(5, WeirdType{50}));
+
+		Verify(Method(mock, funcWeirdType).Using(1, Any())).Once();
+		Verify(Method(mock, funcWeirdType).Using(5, _)).Once();
 	}
 
     void pass_reference_by_value() {

